src/js/App/GlobalFilter/GlobalFilterMenu.js

```diff
@@ -0,0 +1,169 @@
+import React, { useMemo, useState } from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import { TextInput, MenuList, MenuItem, Select, SelectVariant, MenuGroup, Checkbox } from '@patternfly/react-core';
+
+import './global-filter-menu.scss';
+
+export const groupType = {
+  checkbox: 'checkbox',
+  radio: 'radio',
+  button: 'button',
+  plain: 'plain',
+};
+
+const getMenuItems = (groups, filterValueRegex, onChange, calculateSelected) => {
+  const result = groups.map(({ value, label, id, type, items, ...group }) => ({
+    label,
+    value,
+    items: items
+      .filter(
+        (item) =>
+          group.noFilter ||
+          (value && filterValueRegex.test(value)) ||
+          (label && filterValueRegex.test(label)) ||
+          (item.value && filterValueRegex.test(item.value)) ||
+          (item.label && filterValueRegex.test(item.label))
+      )
+      .map((item, index) => ({
+        ...item,
+        key: item.id || item.value || index,
+        value: String(item.value || item.id || index),
+        onClick: (event) => {
+          onChange(
+            event,
+            calculateSelected(type, value, item.value),
+            {
+              value,
+              label,
+              id,
+              type,
+              items,
+              ...group,
+            },
+            item,
+            value,
+            item.value
+          );
+        },
+      })),
+  }));
+  return result.filter(({ noFilter, items = [] }) => noFilter || items.length > 0);
+};
+
+const GlobalFilterMenu = (props) => {
+  const { filterBy, onFilter, groups = [], onChange, selectedTags } = props;
+  const [isOpen, setIsOpen] = useState(false);
+
+  const calculateSelected = (type, groupKey, itemKey) => {
```

Comment:
I'd move this function out of `GlobalFilterMenu` component, this way it's recreated everytime new render cycle happens. I know you have to remeber current `selectedTags` so how about wrapping it with another function.

```JS
const calculateSelected = (selectedTags) => (type, groupKey, itemKey) => {
    const activeGroup = selectedTags[groupKey];
    if (activeGroup) {
      if (type !== groupType.radio && (activeGroup[itemKey] instanceof Object ? activeGroup[itemKey].isSelected : Boolean(activeGroup[itemKey]))) {
        return {
          ...selectedTags,
          [groupKey]: {
            ...(activeGroup || {}),
            [itemKey]: false,
          },
        };
      }

      return {
        ...selectedTags,
        [groupKey]: {
          ...(type !== groupType.radio ? activeGroup || {} : {}),
          [itemKey]: true,
        },
      };
    }

    return {
      ...selectedTags,
      [groupKey]: {
        [itemKey]: true,
      },
    };
};
```

Reply:
Well, this will re-create the function as well. It's just gonna be wrapped. I think using `useCallback` is probably the best option here.

Reply:
The `useCallback` won't help us much in here as we'd have to write complex update function for the `selectedTags`, but well at least the function can be outside of the component making the component more readable.

Reply:
> to write complex update function 

Why? The only mutable variable is the `selectedTags`. We add it as the effect argument and it should be just fine. The rest of the attributes comes as a function args

Reply:
Mhh, we are copying them to new object, so the react should re-calculate the function on each change of `selectedTags` so it should be fine.